### DISCUSSION TITLE
Update state

### DISCUSSION THREAD

**Discussion by heylisp** [mdbergmann/cl-gserver#42] [Q&A]
Hi,
does updating self state inside the :receive lambda involve messaging?
I'm asking because I would like to do it using (setf state ...).
I see that the actor processes the message but the state sometimes is updated and other times not.
Thanks

**Comment by mdbergmann** [comment]
Hi.
Let me ask, why do you want to update the state manually?

**Reply by heylisp** [reply]
I want the actor to change its state based on what message it receives.

**Reply by mdbergmann** [reply]
OK. But there is no need to `setf` the state manually.
Just return with `(cons reply new-state)`. If you have no reply `reply` can be `nil`.

**Reply by mdbergmann** [reply]
Then `new-state` will be the new state of the actor. On a subsequent 'receive' `new-state` (from previous) should eq/equals `state`.

**Reply by heylisp** [reply]
Ok. I did that but the issue seems to remain:
I send a message to one actor, then it sends a message to each actor that should update its state. Some actors are correctly updated, others not. And it changes apparently randomly who is updated and who is not. Sometimes they are all correctly updated.
That's why I supposed that state is updated via messaging.

**Reply by heylisp** [reply]
I run (ac:all-actors *sys*) to check the state of each actor

**Reply by mdbergmann** [reply]
Can you post a snippet of how that works?

One thing to keep in mind is, each actor should have a different instance of state, because otherwise one actor changing state also changes the state of another actor because they share the same state instance. That's of course not wanted.

**Reply by mdbergmann** [reply]
> That's why I supposed that state is updated via messaging.

No. State is not updated via messaging. It's updated as part of the 'receive' function result processing where the actor is in a locked state, so no other messages can interfere.

**Reply by heylisp** [reply]
Finally I figured out the problem: you were right, I messed up with the state. Thank you!
